clang/lib/CodeGen/Targets/NVPTX.cpp

```diff
@@ -285,6 +289,20 @@ void NVPTXTargetCodeGenInfo::addNVVMMetadata(llvm::GlobalValue *GV,
 bool NVPTXTargetCodeGenInfo::shouldEmitStaticExternCAliases() const {
   return false;
 }
+
+llvm::Constant *
+NVPTXTargetCodeGenInfo::getNullPointer(const CodeGen::CodeGenModule &CGM,
+                                       llvm::PointerType *PT,
+                                       QualType QT) const {
+  auto &Ctx = CGM.getContext();
+  if (PT->getAddressSpace() != Ctx.getTargetAddressSpace(LangAS::opencl_local))
+    return llvm::ConstantPointerNull::get(PT);
+
+  auto NPT = llvm::PointerType::get(
+      PT->getContext(), Ctx.getTargetAddressSpace(LangAS::opencl_generic));
+  return llvm::ConstantExpr::getAddrSpaceCast(
+      llvm::ConstantPointerNull::get(NPT), PT);
+}
```

Comment:
I don't quite understand what's going on here. Why are we ASC'ing *all* null pointers to `LangAS::opencl_generic` ?

Will it work for CUDA (as in the CUDA language)? I think this code should be restricted to apply the ASC only for OpenCL and leave CUDA/HIP with the dafault.

Reply:
Below code in the AMDGPU file shows that the ASC is already in place for HIP.
`
 LOC = nullptr;
  // CHECK-DAG: store ptr addrspace(3) addrspacecast (ptr null to ptr addrspace(3)), ptr [[LOC]].ascast, align 4
  `
  For the CUDA case, not ASCing the nullptr in local_space results in undefined behaviour as only the 24 LSBs of the `nullptr` value is 0 and the rest is changing with each run. i.e. if I copy local_space nullptr in the kernel to a long  var and print the value, it won't be zero. So, overall, the ASCing the local_space nullptr for CUDA is needed as it's already there for AMDGPU.

Reply:
Hi @Artem-B 

I'm shimming in at @mmoadeli's request. I advised him on the resolution of his issue.

> I don't quite understand what's going on here.

So it is a similar story as for the AMDGPU backend. `0` as a pointer to shared memory is a valid one and points to the root of the shared memory, so that's means we cannot use this value as `nullptr`. AMDGPU uses -1 (all bits set) for this, but we couldn't find anything equivalent in the CUDA/PTX documentation. After a few investigation, we found out the most stable way to do this is simply by inserting this expression.

Note that `0` as a pointer to the generic address space the expected value for `nullptr`

> Why are we ASC'ing all null pointers to LangAS::opencl_generic ?

The patch isn't doing this, if the pointer type *is* to the cuda shared address space (opencl's local address space) then we do ASC. Otherwise this emits the simple `llvm::ConstantPointerNull`.
We used `LangAS::opencl_generic` as a way to emphasis there is a generic to shared address space cast going on. The other solution here would be to use `LangAS::Default` to retrieve the target address space, but `Default` doesn't sound right to me as you have to know this maps to NVPTX's generic target address space. Either way, we don't have a strong opinion on what to use. But a comment is probably needed regardless.

> Will it work for CUDA (as in the CUDA language)? I think this code should be restricted to apply the ASC only for OpenCL and leave CUDA/HIP with the dafault.

So yes and no. To the `Will it work for CUDA ?` part, yes it will because you actually cannot hit this return. CUDA doesn't expose address spaces so you can't have that nullptr as an address in the cuda shared address space, so the `if` above will always evaluate to true in CUDA.

For the `leave CUDA/HIP with the dafault` part, you could force things and use target address spaces like it is done in the clang headers for CUDA and this change would capture that. However, as explained before, `0` in the address space 3 (NVPTX backend) is a valid address and it is very easy to highlight in SASS.

Reply:
As far as I can tell the reason for the AMDGPU code using an addrspacecast is the following comment `// Currently LLVM assumes null pointers always have value 0, which results in incorrectly transformed IR.` so it can't use a `null` literal for all ones.

Looking at the lang-ref I can't actually see anywhere that `ptr addrspace(X) null` is the zero value, so this should probably be clarified in the lagref: https://llvm.org/docs/LangRef.html#t-pointer

Reply:
"The semantics of non-zero address spaces are target-specific. Memory access through a non-dereferenceable pointer is undefined behavior in any address space. Pointers with the bit-value 0 are only assumed to be non-dereferenceable in address space 0, unless the function is marked with the null_pointer_is_valid attribute."

Reply:
I think this really needs to be a property of the datalayout. the addrspacecast is a roundabout way of using -1

Reply:
Agreed. I believe that right now some LLVM passes assume a `null` constant is all zeroes, even though LangRef does not specify it so we need to either document those semantics or add a property to the data layout.